pkg/inventory/inventory-client.go

```diff
@@ -24,455 +23,304 @@ import (
-		}
-		inv, err := cic.InventoryFactoryFunc(invObj)
-		if err != nil {
-			return nil, err
-		}
-		if err := inv.Store(objs, status); err != nil {
-			return nil, err
-		}
-		klog.V(4).Infof("creating initial inventory object with %d objects", len(objs))
+type WriteClient interface {
+	CreateOrUpdate(ctx context.Context, inv Inventory, opts UpdateOptions) error
+	UpdateStatus(ctx context.Context, inv Inventory, opts UpdateOptions) error
+	Delete(ctx context.Context, inv Info, opts DeleteOptions) error
+}
 
-		if dryRun.ClientOrServerDryRun() {
-			klog.V(4).Infof("dry-run create inventory object: not created")
-			return nil, nil
-		}
+type UpdateOptions struct{}
 
-		err = inv.Apply(ctx, cic.dc, cic.mapper, cic.statusPolicy)
-		return nil, err
-	}
+type GetOptions struct{}
 
-	// Update existing cluster inventory with merged union of objects
-	clusterObjs, err := cic.GetClusterObjs(ctx, localInv)
-	if err != nil {
-		return pruneIDs, err
-	}
-	pruneIDs = clusterObjs.Diff(objs)
-	unionObjs := clusterObjs.Union(objs)
-	var status []actuation.ObjectStatus
-	if cic.statusPolicy == StatusPolicyAll {
-		status = getObjStatus(pruneIDs, unionObjs)
-	}
-	klog.V(4).Infof("num objects to prune: %d", len(pruneIDs))
-	klog.V(4).Infof("num merged objects to store in inventory: %d", len(unionObjs))
-	wrappedInv, err := cic.InventoryFactoryFunc(clusterInv)
-	if err != nil {
-		return pruneIDs, err
-	}
-	if err = wrappedInv.Store(unionObjs, status); err != nil {
-		return pruneIDs, err
-	}
+type ListOptions struct{}
 
-	// Update not required when all objects in inventory are the same and
-	// status does not need to be updated. If status is stored, always update the
-	// inventory to store the latest status.
-	if objs.Equal(clusterObjs) && cic.statusPolicy == StatusPolicyNone {
-		return pruneIDs, nil
-	}
+type DeleteOptions struct{}
 
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run create inventory object: not created")
-		return pruneIDs, nil
-	}
-	err = wrappedInv.Apply(ctx, cic.dc, cic.mapper, cic.statusPolicy)
-	return pruneIDs, err
-}
+var _ Client = &UnstructuredClient{}
 
-// Replace stores the passed objects in the cluster inventory object, or
-// an error if one occurred.
-func (cic *ClusterClient) Replace(ctx context.Context, localInv Info, objs object.ObjMetadataSet, status []actuation.ObjectStatus,
-	dryRun common.DryRunStrategy) error {
-	// Skip entire function for dry-run.
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infoln("dry-run replace inventory object: not applied")
-		return nil
-	}
-	clusterInv, err := cic.getClusterInventoryInfo(ctx, localInv)
-	if err != nil {
-		return fmt.Errorf("failed to read inventory from cluster: %w", err)
-	}
+// UnstructuredInventory implements Inventory while also tracking the actual
+// KRM object from the cluster. This enables the client to update the
+// same object and utilize resourceVersion checks.
+type UnstructuredInventory struct {
+	BaseInventory
```

Comment:
Should this have a pointer or does it need to be pointer-less in order for *UnstructuredInventory to satisfy the interface?